packages/payment-processor/src/payment/erc20-fee-proxy.ts

```diff
@@ -44,49 +46,105 @@ export async function payErc20FeeProxyRequest(
+  request: ClientTypes.IRequestData,
+  signerOrProvider: Web3Provider | Signer = getProvider(),
+  swapSettings: ISwapSettings,
+  amount?: BigNumberish,
+  feeAmount?: BigNumberish,
+  overrides?: ITransactionOverrides,
+): Promise<ContractTransaction> {
+  const proxyAddress = erc20SwapToPayArtifact.getAddress(request.currencyInfo.network!);
+  const signer = getSigner(signerOrProvider);
+
+  const encodedTx = encodePayErc20FeeRequest(request, signerOrProvider, amount, feeAmount, swapSettings);
+
+  const tx = await signer.sendTransaction({
+    data: encodedTx,
+    to: proxyAddress,
+    value: 0,
+    ...overrides,
+  });
+  return tx;
+}
+
 /**
  * Encodes the call to pay a request through the ERC20 fee proxy contract, can be used with a Multisig contract.
  * @param request request to pay
  * @param signerOrProvider the Web3 provider, or signer. Defaults to window.ethereum.
  * @param amount optionally, the amount to pay. Defaults to remaining amount of the request.
  * @param feeAmountOverride optionally, the fee amount to pay. Defaults to the fee amount of the request.
+ * @param swapSettings settings for the swap:
  */
 export function encodePayErc20FeeRequest(
   request: ClientTypes.IRequestData,
   signerOrProvider: Web3Provider | Signer = getProvider(),
   amount?: BigNumberish,
   feeAmountOverride?: BigNumberish,
+  swapSettings?: ISwapSettings,
 ): string {
   validateRequest(request, PaymentTypes.PAYMENT_NETWORK_ID.ERC20_FEE_PROXY_CONTRACT);
-  const signer = getSigner(signerOrProvider);
 
+  const signer = getSigner(signerOrProvider);
   const tokenAddress = request.currencyInfo.value;
-  const proxyAddress = erc20FeeProxyArtifact.getAddress(request.currencyInfo.network!);
-
   const { paymentReference, paymentAddress, feeAddress, feeAmount } = getRequestPaymentValues(
     request,
   );
+  const amountToPay = getAmountToPay(request, amount);
+  const feeToPay = bigNumberify(feeAmountOverride || feeAmount || 0);
 
   if (!!feeAmount !== !!feeAddress) {
     throw new Error('Both fee address and fee amount have to be declared, or both left empty');
   }
-
-  const amountToPay = getAmountToPay(request, amount);
-  const feeToPay = bigNumberify(feeAmountOverride || feeAmount || 0);
-
   if (amountToPay.isZero() && feeToPay.isZero()) {
     throw new Error('Request payment amount and fee are 0');
   }
 
-  const proxyContract = Erc20FeeProxyContract.connect(proxyAddress, signer);
+  if (!swapSettings) {
+    // Payment with the request currency
+    const proxyAddress = erc20FeeProxyArtifact.getAddress(request.currencyInfo.network!);
+    const proxyContract = Erc20FeeProxyContract.connect(proxyAddress, signer);
+
+    return proxyContract.interface.functions.transferFromWithReferenceAndFee.encode([
+      tokenAddress,
+      paymentAddress,
+      amountToPay,
+      `0x${paymentReference}`,
+      feeToPay,
+      feeAddress || constants.AddressZero,
+    ]);
+  }
+
+  // Swap to pay
```

Comment:
This makes this function a lot more complex.
I would suggest creating another module for the swap payments.
Something like `erc20-swap-payment.ts`, and having all the swap logic there, if possible.

Changes like this one, `payErc20Request()`, `approveErc20ForSwapToPayIfNeeded()`, `approveErc20ForSwapToPay()`, the interface `ISwapSettings`, and all things that concern only swapping should be encapsulated on it's own module as much as possible.